tutorial.cxx

```diff
@@ -653,7 +654,81 @@ void f32_matrix_multiplication_4x4_loop_avx512_kernel(float a[4][4], float b[4][
 
     _mm512_storeu_ps(&c[0][0], c_mat);
 }
+*/
 
+void f32_matrix_multiplication_4x4_loop_avx512_kernel(float a[4][4], float b[4][4], float c[4][4]) {
+    __m512 a_mat = _mm512_loadu_ps(&a[0][0]);
+    __m512 b_mat = _mm512_loadu_ps(&b[0][0]);
+
+    __m512i trans_perm = _mm512_setr_epi32( //
+        0, 4, 8, 12,                        //
+        1, 5, 9, 13,                        //
+        2, 6, 10, 14,                       //
+        3, 7, 11, 15                        //
+    );
+    
+    __m512i lshiftz_8_perm = _mm512_setr_epi32( //
+        8, 9, 10, 11,                           //
+        12, 13, 14, 15,                         //
+        0, 0, 0, 0,                             //
+        0, 0, 0, 0                              //
+    );
+    
+    __m512i rot_4_perm = _mm512_setr_epi32( //
+        4, 5, 6, 7,                         //
+        8, 9, 10, 11,                       //
+        12, 13, 14, 15,                     //
+        0, 1, 2, 3                          //
+    );
+
+    // first C row
+    __m512 a_row_1_broad = _mm512_broadcast_f32x4(_mm512_castps512_ps128(a_mat));
+    __m512 a_row_1_broad_trans = _mm512_permutexvar_ps(trans_perm, a_row_1_broad);
+    __m512 c_row_1 = _mm512_setzero_ps();
+    c_row_1 = _mm512_fmadd_ps(a_row_1_broad_trans, b_mat, c_row_1);
+    __m512 c_row_1_rot = _mm512_permutexvar_ps(lshiftz_8_perm, c_row_1);
+    c_row_1 = _mm512_add_ps(c_row_1, c_row_1_rot);
+    c_row_1_rot = _mm512_permutexvar_ps(rot_4_perm, c_row_1);
+    c_row_1 = _mm512_add_ps(c_row_1, c_row_1_rot);
+    _mm512_mask_compressstoreu_ps(&c[0][0], 0xF, c_row_1);
```

Comment:
We should be able to perform this entire kernel with just 2 loads, 1 store. In your case, there are 4 stores.